### PR TITLE
Make models and cli subpackages pyflakes compliant

### PR DESCRIPTION
**Pull request by almarklein**
issues: ref #2078 

These changes make the `bokeh.cli` and `bokeh.models` subpackages compliant with pyflakes. Especially the latter needed significant changes in the `__init__`; pyflakes does not allow `from x import *`, so I had to make them all explicit imports. But the list of names to import became so long that I wrote a small function to collect these names. The names to import is particularly long for the widgets module.

To shut pyflakes up about unused imports, apparently you can also use `__all__` which I personally found a bit nicer than just stating the names.  Happy to revert this to just stating a tuple of names though.

I am generally in favor of explicit imports, but in the case of importing all widgets, a star-import would be justified IMO. However, there is no way to shut pyflakes up unless we use flake8 ...

### CODE REVIEW

File: bokeh/models/__init__.py [hunk 1]
```diff
@@ -1,18 +1,96 @@
...
+from .widgets import (AbstractButton, AbstractGroup, AbstractIcon, AutocompleteInput,
+                      BaseBox, BooleanFormatter, Button, ButtonGroup, CellEditor,
+                      CellFormatter, CheckboxButtonGroup, CheckboxEditor,
+                      CheckboxGroup, DataTable, DateEditor, DateFormatter, DatePicker,
+                       DateRangeSlider, Dialog, Dropdown, Group, HBox, Icon,
+                       InputWidget, IntEditor, Layout, MultiSelect, NumberEditor,
+                       NumberFormatter, Panel, Paragraph, PercentEditor, PreText,
+                       RadioButtonGroup, RadioGroup, Select, SelectEditor, Slider,
+                       StringEditor, StringFormatter, TableColumn, TableWidget, Tabs,
+                       TextEditor, TextInput, TimeEditor, Toggle, VBox, VBoxForm)
```
Comment (almarklein) [inline bokeh/models/__init__.py]: This is where I started to feel rather uncomfortable ...

**Comment by bryevdv** [comment]
Honestly, I don't think this is an improvement, regardless of what pyflakes says. I don't think we should do this just to appease pyflakes. If we can't silence individual checks, what about excluding entire modules from the run? We could just leave `bokeh/models/__init__.py` out.

**Comment by bryevdv** [comment]
Options I see:
* Don't import everything in to `bokeh.models` at all. It seems like a nice convenience though I believe @mattpap would prefer this. 
* Automate all the imports in a way that satisfies pyflakes, but does not require listing them either last things are variables.

**Comment by almarklein** [comment]
> Honestly, I don't think this is an improvement, regardless of what pyflakes says.

I agree with that :) Problem is that vanilla pyflakes has no way to ignore either individual lines or individual files. I've seen some solutions that involve monkey-patching pyflakes, but I'd rather use flake8 in that case.

One option could be something along these lines:

    for submodule in ['actions', 'axes', 'formatters', 'and', 'other', 'submodules']:
        for name, ob in import_sub_module_via__import__(submodule):
             if name.startswith('_'):
                continue
             if ob.__module__.startswith('bokeh.models'):
                 globals()[name] = ob
 
It would basically `import *` from all listed submodules, but only for the classes that are actually defined there (e.g. not Int, String, whatever that were imported in the submodule).

**Comment by bryevdv** [comment]
I'm OK with that.

**Comment by mattpap** [comment]
You can use `__all__` instead in submodules and then concatenate them in `__init__`. `__all__` is insane on its own, but you can use my `@public` decorator [1] to avoid repetition. 

[1] https://github.com/sympy/sympy/blob/master/sympy/utilities/decorator.py#L147

**Comment by bryevdv** [comment]
That seems pretty useful.

**Comment by mattpap** [comment]
> Do not automate pyflakes checking all (reduce warnings through periodic, scheduled manual dev checks?)

That seem about right. Thought, some automation is great, e.g. checking for mixed tabs and spaces, especially if there are going to be inexperienced contributors.

**Comment by almarklein** [comment]
> You can use __all__ instead in submodules and then concatenate them in __init__

Unfortunately, that won't work, since pyflakes is a static analysis tool; it does not execute any of the code. Even in this small example pyflakes will say "'foo imported but unused'":

    from x import foo
    __all__ = []
    __all__.append('foo')

> Do not automate pyflakes checking all (reduce warnings through periodic, scheduled manual dev checks?)

Even if we would do this, these periodic tests would be painful if they are riddled with false positives.

**Comment by bryevdv** [comment]
Well, we can consider `flake8` but it would require some tuning to for some of our style specifics.

**Comment by almarklein** [comment]
> Well, we can consider flake8 but it would require some tuning to for some of our style specifics. 

That should not be too hard; we can start with just turning everything off except pyflakes stuff. Afaik this stuff can be specified in a config file, so one can simply run `flake8` and it would work as intended.

The solution (auto-importing from submodules using `imp` or `__import__` is not very pretty, but prettier than what we have now, and allows vanilla pyflakes. Personally, I feel awkward jumping through hoops like this just because we cannot shut up pyflakes, so I think I lean towards using flake8. But its your call :)

**Comment by bryevdv** [comment]
Personally I'd rather leave the `*` imports as they are and investigate `flake8`

**Comment by almarklein** [comment]
Ok, I will modify this PR accordingly, so we can see what that will look like ...

File: setup.cfg [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+# Style checks turned on:
+#   F - all pyflake errors
+#   E101 - indentation contains mixed spaces and tabs
+#   E111 - indentation is not a multiple of four
+#   E501 - line too long (see max-line-length)
+
+exclude = bokeh/models/__init__.py, bokeh/models/widgets/__init__.py
+ignore = E,W
+select = F,E101,E111,E501
+max-line-length = 120
```
Comment (almarklein) [inline setup.cfg]: This is where we can configure flake8. This turns off all style errors and warnings, and then turns on just the ones we want. I think E101 and E111 make sense. E501 can be turned off if majority prefers that; I'm just showing what we can do (it now allows lines up to 120 chars)

**Comment by almarklein** [comment]
Ok, I changed the code to use flake8. The `__init__.py` in models and widgets are both excluded from any kind of checking. I kept the explicit imports, to keep the namespace clean. if people prefer to revert to `from .x import *` I'm happy to change that (although I personally find it less pretty :) ).

**Comment by bryevdv** [comment]
In this case I would actually prefer that. It's an unnecessary maintenance burden to have to remember to put new or changed widgets in the list. The point is to import them all, so just import them all, I'd say.

**Comment by almarklein** [comment]
Will do.

Upon inspecting the travis error, I see that in some places we do things like `from ..models import ServerDataSource,  GlyphRenderer, Range1d, Color`. It might be cleaner to get `Range1d` and `Color` from where they are defined, or is the models namespace ok for this?

Anyway, its a minor detail: these errors should disappear if I put the `*` back.

**Comment by bryevdv** [comment]
In the library code I agree it would be better to import from the actual module where the thing is defined. `bokeh.models` is really a convenience for users.

File: setup.cfg (hunk 2 above)
Comment (almarklein) [inline setup.cfg]: @bryevdv what's your opinion about imposing a max line length? We can always change/remove it of course, so maybe we can try imposing a relatively generous width to pick out the worst cases?

Comment (bryevdv) [inline setup.cfg]: I'm relatively OK with trying that out.

**Comment by almarklein** [comment]
Ready to go then :)

**Comment by bryevdv** [comment]
LGTM, merging

**Pull request merged**